src/device-api/common.js

```diff
@@ -62,9 +62,55 @@ export async function doPurge(appId, force) {
 			const currentServices = app.services;
 			const currentVolumes = app.volumes;
 
+			/**
+			 * With multi-container, Docker adds an invalid network alias equal to the current containerId
+			 * to that service's network configs when starting a service. Thus when reapplying intermediateState
+			 * after purging, filter out invalid network aliases from service config objects.
+			 * This will prevent error logs like the following:
+			 *
+			 * https://gist.github.com/cywang117/84f9cd4e6a9641dbed530c94e1172f1d#file-logs-sh-L58
+			 *
+			 * When networks do not match because of their aliases, services are killed and recreated
+			 * an additional time which is unnecessary. Filtering prevents this additional restart BUT
+			 * it is a stopgap measure until common.js is typed.
+			 */
+			currentServices.forEach((svc) => {
+				if (svc.config.networks) {
+					Object.keys(svc.config.networks).forEach((network) => {
+						// If it's Supervisor's network, the only alias should be a containerId which we
+						// don't want as a network alias anyway (see reasons above), so we can remove the entire array
+						if (network.includes('supervisor')) {
+							svc.config.networks[network].aliases = [];
+						} else {
+							// We know that svc.serviceName shouldn't be null in this context,
+							// but we can't add any typings since common.js is untyped
+							svc.config.networks[network].aliases = [
+								typeof svc.serviceName === 'string' ? svc.serviceName : '',
+							];
+						}
+					});
+				}
+			});
+
+			/**
+			 * See https://github.com/balena-os/balena-supervisor/blob/master/src/device-api/common.js#L160-L180
+			 * for a more in-depth explanation of why aliases need to be filtered out.
+			 *
+			 * NOTE that we can't use safeStateClone which performs this filtering function because common.js
+			 * is not yet written in TypeScript, leading to a type error when trying to pass cloned state into
+			 * the 2nd applyIntermediate call.
+			 *
+			 * TODOs in order of achievability:
+			 * - Convert common.js to TypeScript & switch to using safeStateClone
+			 * - Prevent network aliases that are containerIds from being added to service config objects
+			 */
```

Comment:
@20k-ultra @pipex Here is the problem area I mentioned in my other comment. Hopefully I've documented in well enough here but lemme know if there are any questions. The `currentServices.forEach` block to filter out containerId aliases definitely feels like a bit of a hack, but in the absence of typed common.ts I'm not sure what to do to prevent `Networks do not match!` issue.

Reply:
Idea: don't merge this PR until `common.js` is converted? I could work on that next and the file isn't too big so it shouldn't take long (kappa).

Reply:
With the [latest changes](https://github.com/balena-os/balena-supervisor/commit/bdf8808c716b19b866c2c240e5dddf90918206f4) courtesy of @Page- , please ignore this comment about holding off merge til common.ts.